File: rangecoder.py
import time, sys, getopt, os.path

ENCODE = "encode"
DECODE = "decode"
MAX_RANGE = 0x100000000
MIN_RANGE = 0x1000000
MASK      = 0xffffffff
SHIFT     = 24

# バイト単位の入出力
def getc(f):
    c = f.read(1)
    if c == b'': return None
    #return ord(c)
    return int.from_bytes(c, "big")

def putc(f, x):
    #f.write(chr(x & 0xff))
    f.write(x.to_bytes(1, "big"))

#
class RangeCoder:
    def __init__(self, file, mode):
        self.file = file
        self.range = MAX_RANGE
        self.buff = 0
        self.cnt = 0
        if mode == ENCODE:
            self.low = 0
        elif mode == DECODE:
            # buff の初期値 (0) を読み捨てる
            getc(self.file)
            # 4 byte read
            self.low = getc(self.file)
            self.low = (self.low << 8) + getc(self.file)
            self.low = (self.low << 8) + getc(self.file)
            self.low = (self.low << 8) + getc(self.file)
        else:
            raise "RangeCoder mode error"

    # 符号化の正規化
    def encode_normalize(self):
        if self.low >= MAX_RANGE:
            # 桁上がり
            self.buff += 1
            self.low &= MASK
            if self.cnt > 0:
                putc(self.file, self.buff)
                for _ in range(self.cnt - 1): putc(self.file, 0)
                self.buff = 0
                self.cnt = 0
        while self.range < MIN_RANGE:
            if self.low < (0xff << SHIFT):
                putc(self.file, self.buff)
                for _ in range(self.cnt): putc(self.file, 0xff)
                self.buff = (self.low >> SHIFT) & 0xff
                self.cnt = 0
            else:
                self.cnt += 1
            self.low = (self.low << 8) & MASK
            self.range <<= 8

    # 復号の正規化
    def decode_normalize(self):
        while self.range < MIN_RANGE:
            self.range <<= 8
            c = getc(self.file)
            if c == None: 
                self.low = ((self.low << 8) + 0) & MASK
                return
            self.low = ((self.low << 8) + c) & MASK

    # 終了
    def finish(self):
        c = 0xff
        if self.low >= MAX_RANGE:
            # 桁上がり
            self.buff += 1
            c = 0
        putc(self.file, self.buff)
        for _ in range(self.cnt): putc(self.file, c)
        #
        putc(self.file, (self.low >> 24) & 0xff)
        putc(self.file, (self.low >> 16) & 0xff)
        putc(self.file, (self.low >> 8) & 0xff)
        putc(self.file, self.low & 0xff)


# 出現頻度表
class Freq:
    def __init__(self, count):
        size = len(count)
        self.size = size
        m = max(count)
        # 2 バイトに収める
        if m > 0xffff:
            self.count = [0] * size
            n = 0
            while m > 0xffff:
                m >>= 1
                n += 1
            for x in range(size):
                if count[x] != 0:
                    self.count[x] = (count[x] >> n) | 1
        else:
            self.count = count[:]
        self.count_sum = [0] * (size + 1)
        # 累積度数表
        for x in range(size):
            self.count_sum[x + 1] = self.count_sum[x] + self.count[x]

    #
    def write_count_table(self, fout):
        for x in self.count:
            putc(fout, x >> 8)
            putc(fout, x & 0xff)
    
    # 符号化
    def encode(self, rc, c):
        temp = int(rc.range / self.count_sum[self.size])
        rc.low += self.count_sum[c] * temp
        rc.range = self.count[c] * temp
        rc.encode_normalize()

    # 復号
    def decode(self, rc):
        # 記号の探索
        def search_code(value):
            i = 0
            j = self.size - 1
            while i < j:
                k = int((i + j) / 2)
                if self.count_sum[k + 1] <= value:
                    i = k + 1
                else:
                    j = k
            return i
        #
        temp = int(rc.range / self.count_sum[self.size])
        c = search_code(rc.low / temp)
        rc.low -= temp * self.count_sum[c]
        rc.range = temp * self.count[c]
        rc.decode_normalize()
        return c

# ファイルの読み込み
def read_file(fin):
    while True:
        c = getc(fin)
        if c is None: break
        yield c

# レンジコーダによる符号化
def encode(fin, fout):
    count = [0] * 256
    for x in read_file(fin):
        count[x] += 1
    rc = RangeCoder(fout, ENCODE)
    freq = Freq(count)
    freq.write_count_table(fout)
    fin.seek(0)
    for x in read_file(fin):
        freq.encode(rc, x)
    rc.finish()

# 出現頻度表の読み込み
def read_count_table(fin):
    count = [0] * 256
    for x in range(256):
        count[x] = (getc(fin) << 8) + getc(fin)
    return count

# レンジコーダによる復号
def decode(fin, fout, size):
    freq = Freq(read_count_table(fin))
    rc = RangeCoder(fin, DECODE)
    for _ in range(size):
        putc(fout, freq.decode(rc))

# 符号化
def encode_file(name1, name2):
    size = os.path.getsize(name1)
    infile = open(name1, "rb")
    outfile = open(name2, "wb")
    putc(outfile, (size >> 24) & 0xff)
    putc(outfile, (size >> 16) & 0xff)
    putc(outfile, (size >> 8) & 0xff)
    putc(outfile, size & 0xff)
    print(size)
    if size > 0: encode(infile, outfile)
    infile.close()
    outfile.close()

# 復号
def decode_file(name1, name2):
    infile = open(name1, "rb")
    outfile = open(name2, "wb")
    size = 0
    for _ in range(4):
        size = (size << 8) + getc(infile)
    print(size)
    if size > 0: decode(infile, outfile, size)
    infile.close()
    outfile.close()

#
def main():
    eflag = False
    dflag = False
    opts, args = getopt.getopt(sys.argv[1:], 'ed')
    for x, y in opts:
        if x == '-e' or x == '-E':
            eflag = True
        elif x == '-d' or x == '-D':
            dflag = True
    if eflag and dflag:
        print('option error')
    elif eflag:
        encode_file(args[0], args[1])
    elif dflag:
        decode_file(args[0], args[1])
    else:
        print('option error')

#
s = time.clock()
main()
e = time.clock()
print("%.3f" % (e - s))

File: go/rangecoder.go
package main

import (
	"fmt"
	"os"
	"encoding/binary"
)

const MAX_RANGE uint32 = 0xffffffff
const MIN_RANGE uint32 = MAX_RANGE / 256
const MASK      uint32 = 0xfffffff
const SHIFT uint = 24

func main(){
	flag := os.Args[1]
	input := os.Args[2]
	output := os.Args[3]

	if flag == "encode" {
		encode_file(input, output)
	}else if flag == "decode" {
		decode_file(input, output)
	}
}


//---------------------------------------------------------
// encode
//---------------------------------------------------------
func encode_file(input, output string){
	size, array := readBinary(input)

	// sub
	//array2 := make([]uint16, len(array))
	//array2[0] = array[0] + 1024
	//for i := 1; i < len(array); i++ {
	//	array2[i] = array[i] - array[i - 1] + 1024
	//}

	out, _ := os.Create(output)
	//defer out.Close()
	binary.Write(out, binary.LittleEndian, size / 2)

	fmt.Println(size)
	encode(array, out)
}

func encode(array []uint16, out *os.File){
	count := count_table(array)

	encoder := &Encoder{}
	encoder.init(out)
	encoder.init_count_table(count)
	encoder.write_count_table()

	for _, x := range array {
		encoder.encode(x)
	}
	encoder.finish()
}

func count_table(array []uint16) ([]uint32){
	count := make([]uint32, 256 * 256)
	for _, x := range array {
		count[x] += 1
	}
	return count
}


type Encoder struct {
	file *os.File
	rng uint32
	low uint32
	buff uint8
	cnt int
	count []uint16
	count_sum []uint32
	count_all float64
}

func (p *Encoder) init(file *os.File){
	p.file = file
	p.rng = MAX_RANGE
}

func (p *Encoder) init_count_table(count []uint32){
	var max uint32
	var maxi int
	for i, x := range count {
		if x > max {
			max = x
		}
		if x > 0 {
			maxi = i
		}
	}
	fmt.Println(max, maxi)
	counts := count[:maxi + 1]

	p.count = make([]uint16, len(counts))

	// create uint16 count table
	if max > 0xffff {
		var n uint
		m := max
		for m > 0xffff {
			m >>= 1
			n += 1
		}

		for i, x := range counts {
			if x != 0 {
				p.count[i] = uint16((x >> n) | 1)
			}
		}

	}else{
		for i, x := range counts {
			p.count[i] = uint16(x)
		}
	}

	// sum count
	p.count_sum = make([]uint32, len(counts) + 1)
	for i, x := range p.count {
		p.count_sum[i + 1] = p.count_sum[i] + uint32(x)
	}
	p.count_all = float64(p.count_sum[len(p.count_sum) - 1])

	if p.count_sum[len(p.count_sum) - 1] > MIN_RANGE {
		fmt.Println("count_all > MIN_RANGE")
	}
}

func (p *Encoder) write_count_table(){
	binary.Write(p.file, binary.LittleEndian, uint16(len(p.count)))
	binary.Write(p.file, binary.LittleEndian, p.count)
}

func (p *Encoder) encode(c uint16){
	temp := float64(p.rng) / p.count_all
	p.low += uint32(float64(p.count_sum[c]) * temp)
	p.rng = uint32(float64(p.count[c]) * temp)
	//if p.rng == 0 {
	//	fmt.Println("zero", p.rng, p.count[c], float64(p.count[c]) * temp)
	//}
	p.normalize()
}
func (p *Encoder) normalize(){
	// carry low
	if p.low >= MAX_RANGE {
		p.buff += 1
		p.low &= MASK

		if p.cnt > 0 {
			p.puti(p.buff)
			for i := 0; i < p.cnt - 1; i++ {
				p.puti(0)
			}
			p.buff = 0
			p.cnt = 0
		}
	}

	// scale range
	for p.rng < MIN_RANGE {
		if p.low < (0xff << SHIFT) {
			p.puti(p.buff)
			for i := 0; i < p.cnt; i++ {
				p.puti(0xff)
			}
			p.buff = uint8(p.low >> SHIFT)
			p.cnt = 0
		}else{
			p.cnt += 1
		}

		p.low = (p.low << 8) & MASK
		p.rng <<= 8
	}
}

func (p *Encoder) finish(){
	var c uint8 = 0xff
	if p.low >= MAX_RANGE {
		p.buff += 1
		c = 0
	}
	p.puti(p.buff)
	for i := 0; i < p.cnt; i++ {
		p.puti(c)
	}

	fmt.Println(p.low)
	binary.Write(p.file, binary.LittleEndian, p.low)
}

func (p *Encoder) puti(i uint8){
	binary.Write(p.file, binary.LittleEndian, i)
}

func readBinary(filename string) (size int64, array []uint16){
	fileinfo, _ := os.Stat(filename)
	size = fileinfo.Size()

	file, _ := os.Open(filename)
	defer file.Close()
	array = make([]uint16, size / 2)
	binary.Read(file, binary.LittleEndian, &array)
	return
}


//---------------------------------------------------------
// decode
//---------------------------------------------------------
func decode_file(input, output string){
	out, _ := os.Create(output)
	defer out.Close()

	decode(out, input)
}

func read_encoded(filename string) (int64, []uint8, []uint16){
	fileinfo, _ := os.Stat(filename)
	filesize := fileinfo.Size()

	file, _ := os.Open(filename)

	var size int64
	binary.Read(file, binary.LittleEndian, &size)

	count_table := read_count_table(file)

	datasize := int(filesize) - 8 - 2 * (len(count_table) + 1)
	fmt.Println(datasize)
	data := make([]uint8, datasize)
	err := binary.Read(file, binary.LittleEndian, &data)
	if err != nil {
		fmt.Println("binary.Read failed:", err)
	}

	return size, data, count_table
}

func read_count_table(file *os.File) (count_table []uint16){
	var maxi uint16
	binary.Read(file, binary.LittleEndian, &maxi)

	count_table = make([]uint16, maxi)
	binary.Read(file, binary.LittleEndian, &count_table)
	return
}

func decode(out *os.File, input string){
	size, data, count_table := read_encoded(input)

	decoder := &Decoder{}
	decoder.init(data)
	decoder.init_count_table(count_table)

	fmt.Println(size)
	for i := 0; i < int(size); i++ {
		code := decoder.decode()
		binary.Write(out, binary.LittleEndian, code)
	}
}

type Decoder struct {
	data []uint8
	readi int
	rng uint32
	low uint32
	count []uint16
	count_sum []uint32
	count_all float64
}

func (p *Decoder) init(data []uint8){
	p.data = data
	p.rng = MAX_RANGE

	p.low = (((((uint32(data[1]) << 8) + uint32(data[2])) << 8) + uint32(data[3])) << 8) + uint32(data[4])
	p.readi = 4
}

func (p *Decoder) init_count_table(count []uint16){
	p.count = count

	// sum count
	p.count_sum = make([]uint32, len(count) + 1)
	for i, x := range p.count {
		p.count_sum[i + 1] = p.count_sum[i] + uint32(x)
	}
	p.count_all = float64(p.count_sum[len(p.count_sum) - 1])
}

func (p *Decoder) decode() uint16 {
	temp := float64(p.rng)  / p.count_all
	code := p.search_code(uint32(float64(p.low) / temp))
	p.low -= uint32(temp * float64(p.count_sum[code]))
	p.rng = uint32(temp * float64(p.count[code]))
	p.normalize()

	return uint16(code)
}

func (p *Decoder) search_code(value uint32) int {
	i := 0
	j := len(p.count) - 1
	for i < j {
		med := int((i + j) / 2)
		if p.count_sum[med + 1] <= value {
			i = med + 1
		}else{
			j = med
		}
	}
	return i
}

func (p *Decoder) normalize() {
	for p.rng < MIN_RANGE {
		p.rng <<= 8
		p.readi += 1
		if p.readi >= len(p.data) {
			p.low = ((p.low << 8) + 0) & MASK
			return

		}else{
			c := p.data[p.readi]
			p.low = ((p.low << 8) + uint32(c)) & MASK
		}
	}
}

